README
"FiSH 10" - a blowfish encryption plug-in for mIRC 7,
  compatible to previous FiSH scripts and other clients!

This is an encryption addon for mIRC, it is based on FiSH v1.30 and is compatible to the
original 'blowcrypt/bloW' script as well as Mircryption and other clients.
It supports private chat, channel, topic encryption and comes with a secure key-exchange
system (DH1080). In addition to the regular ECB Blowfish mode, it also supports the more
secure CBC mode.

Supported mIRC versions: mIRC 7.x only!
	You can download the latest version from www.mirc.com

== INSTALLING ==

	Don't shy away, the guide below is just verbose, installing is easy!

	Setting up FiSH 10 should be pretty straightforward as patching mirc.exe is no longer
	required. If you have been using FiSH.dll before, it is however strongly recommended
	that you restore mirc.exe to its unpatched, original state before you install FiSH 10.
	You can of course apply a "crack" after that, but mIRC is a good product, so if you use
	it every day, I strongly advise you to get a legit lifetime license.

	Alright, now on to the actual installing procedure. If you have not been using FiSH.dll
	before, or if you are installing mIRC 7 from scratch, you can skip this paragraph.
		- Unload FiSH.mrc from mIRC like so: //unload -rs $shortfn($nofile($mircexe)) $+ FiSH.mrc)
		(- Unload blow.mrc, blowcrypt.mrc, mircryption.mrc, etc.)
		- Delete FiSH.dll
		- Delete FiSH.mrc
		(- Delete DH.dll blowfish.dll bloW.dll)
		- Keep! blow.ini

	Now that your mIRC installation is clean, we can move on to the actual installing part.
		- Download the latest zip file from http://github.com/flakes/mirc_fish_10/downloads
			(if you haven't already)
		- Shut down mIRC!
		- Extract fish_10.dll, fish_inject.dll and fish_10.mrc to your mirc.exe folder.
			Do the same with libeay32.dll and ssleay32.dll if your download contains them.
		- If necessary, extract blow.ini-EXAMPLE and rename the file to "blow.ini".
		- You can edit fish_10.mrc if your blow.ini is at a different location than that folder.
			This is useful for people who installed mIRC into Program Files but keep their settings
			in the Users/AppData folder.
		- Start mIRC back up. If your mIRC is automatically connecting on startup, you might
			have to turn that off, or add a timer. It is extremely important that fish_10.mrc
			always loads before ANY IRC connection is made.
		- Install the script: //load -rs $shortfn($nofile($mircexe) $+ fish_10.mrc)
		- Shut down mIRC, and start it again
		- Two lines like this should show up (and no error popup):
			*** FiSH 10 *** by [c&f] *** fish_inject.dll compiled XXX XX 2011 12:00:00 ***
			*** FiSH 10 *** by [c&f] *** fish_10.dll     compiled XXX XX 2011 12:00:00 ***
		- If that is the case, FiSH 10 is now ready for action!
		- Connect to your networks and do your thing.

	If you get an error like "/dll: unable to open file" on startup, install this:
	www.microsoft.com/downloads/details.aspx?familyid=A5C84275-3B97-4AB7-A40D-3802B2AF5FC2
	Don't install the x64 version, even if you have a 64-bit OS. mIRC and FiSH are 32-bit,
	so you need the x86 version!

	If an error like "patching ... failed" shows up, something seems wrong with your setup.
	It could be something simple like an incompatible OpenSSL DLL, something weird like
	your AV scanner, or a bug in FiSH 10. If you made sure you installed the latest version,
	you can check here for existing problem reports, or create a new one:
	http://github.com/flakes/mirc_fish_10/issues
	(Hint for Vista and above: Hit Ctrl + C to copy the message box contents into the clipboard,
	then paste into your bug report using Ctrl + V)

	When mIRC updates, FiSH 10 should just continue to work in 99% of the cases, unlike
	the previous solution. And it's open source, pretty cool, he?

	You can report bugs at http://github.com/flakes/mirc_fish_10/issues
	or ask for help in the old FiSH.dll's forums: http://fish.secure.la/forum/
	Please don't message me on IRC.

	If you are an IRC enthusiast like I am, you should also check out ZNC, the best
	bouncer software available: http://en.znc.in/

	FiSH 10 does not contain any special code for decrypting psyBNC or sBNC logs.
	Why should it? ZNC's buffers play back just fine, including appended (not prepended)
	time stamps. I do consider adding support for psyBNC's network prefix thing at some
	point though, since there seems to be a bunch of people actually using that.

== Multi-network support ==

	The old FiSH never supported using different keys for #chan on NetworkOne, and #chan
	on NetworkTwo. The new FiSH does fully support this while still maintaining blow.ini
	backwards compatibility in all directions.

== CBC MODE ==

	FiSH 10 implements Mircryption's CBC mode, as defined here:
	www.donationcoder.com/Software/Mouser/mircryption/extra_cbcinfo.php

	FiSH 10 is the first add-on to deploy a fully backwards compatible DH1080 key exchange
	that AUTOMATICALLY enables the more secure CBC encryption mode if both parties use
	FiSH 10. You won't notice a difference, but your messages will be encrypted stronger
	than ever. Key exchanges with users of older versions and other scripts are completely
	unimpaired.

	There are however at least two broken DH1080 implementations that take the remainder of
	the string in DH1080_(INIT|FINISH) instead of splitting at space characters:
		- "[G]Script (OrbitIRC)" and "Trillian Astra".
	You can use the right-click menu in queries to selectively disable CBC mode key exchange
	for those users. Please note that this is only a problem if you initiate the key exchange,
	not if they do.

== ABOUT ==

	fish_10.mrc is mostly a 1:1 copy of FiSH.mrc, with quite a number of modifications to
	enable proper Windows Vista and 7 support, to facilitate DLL loading, for CBC mode and
	multi-network awareness.

	I would like to thank RXD for the work he has put into the FiSH addon over the last
	few years. However, he has been slow with updates and unwilling to open source a
	simple mIRC addon like this, so someone had to take over!

	FiSH 10 is using "CPatch" by armagedescu, thanks. It contains some code from
	http://dirtirc.sourceforge.net/ (blowfish core) and some utility methods from
	http://code.google.com/p/infekt/ (which is an excellent NFO viewer by the way).
	The rest of the code is what I contributed myself and is licensed under the
	"Go fuck yourself, it's free!" license.

== TECHNICAL ==

	FiSH 10 makes heavy use of OpenSSL. It is recommended that you use the SSL libs
	(ssleay32.dll and libeay32.dll) that come with the download, but other recent
	libraries should work too. If they don't, FiSH 10 will let you know on startup.

	You can read some lines about the in-memory patching in fish-inject.cpp. fish_inject.dll
	(compiled from fish-inject.cpp, patcher.cpp and socket.cpp) does the magic of intercepting
	the system calls and managing buffers. For complete lines, it calls into fish_10.dll which
	does the actual IRC parsing and encryption things.
	Key-exchange, getting the "outside" IP etc. is still handled by fish_10.mrc, like it
	has always been.

== SECURITY NOTE ==

	You are advised to use TrueCrypt or a similar solution to protect your blow.ini file. While all
	the keys are stored encrypted, the encryption is not particularly strong and could be broken
	without a lot of effort!

	Furthermore, DON'T exchange keys via plaintext (IRC/email/Facebook/whatever). This defeats the
	entire purpose of encryption. Instead, use DH1080 key exchange to establish a secure connection
	for (channel) key exchange.

	And, obviously, DO NOT store any log files on unencrypted disks.

	When using DH1080, please be aware that DH1080 does NOT protect you against sophisticated man-
	in-the-middle attacks where the attacker is able to read and modify your IRC data stream in
	real time.
	You can spot an attacker like that by using the Show key right-click menu entry, and comparing
	the key contents via an absolutely man-in-the-middle-proof channel, such as via telephone or a
	meeting in person.

== KNOWN ISSUES ==

	If you are using a SOCKS or HTTP proxy then FiSH 10 won't work for you. Sorry.

	FiSH 10 is incompatible to the following applications. Please whitelist mirc.exe in these:
		- "AdMuncher"
	This is because FiSH 10 integrates itself into mIRC similarly to these apps.

	If you continue to use your old blow.ini file, and have been using a non-ANSI-compatible crypt
	mark (such as the default middot "·"), using any right click menu from within mIRC will convert
	your blow.ini file from ANSI to UTF-8 which will effectively destroy FiSH 10's ability to
	properly read the crypt mark.
	The suggested workaround is using $chr(183) instead of "·".

	When using psyBNC's multi-network feature (network prefixes), it's VITAL that you define one of
	your networks as "main network". Otherwise, psyBNC won't send a NETWORK name when connecting
	and all DH1080 key exchanges as well as some other things will fail silently.

== THE FUTURE ==

	- A feature that tries to avoid sending truncated fish messages and splits the message into
		several lines automatically (with an option to turn it off).
	- Add support for Micryption's DH1080_INIT_cbc and DH1080_FINISH_cbc method.
	- Make a DH1080 specs document


NEWS
== 2023-07-08 ==
	* Слава Україні! Смерть окупантам!
	* Compatible with mIRC 7.73 and above. No longer compatible with older versions.
	* OpenSSL updated to 3.0.9, now using version built by https://kb.firedaemon.com/support/solutions/articles/4000121705
	* [maroonbells] Enable decryption of notices from group target ~#chan and &#chan with assumption that &#something is not a server-local channel, but is if anything except '#' is the 2nd char
	* [maroonbells] Change do-not-encrypt-NOTICE prefix from DH1080_ to DH | allows scripted change to have a different DH handshake using a shorter prefix
	* [maroonbells] Allow hex-editing alternate g generator

== 2020-10-10 ==
	* Happy New Year
	* OpenSSL updated to 1.1.1h - issue #66
	* Fixed missing encryption of outgoing messages when IRCv3 is used - issue #67
	* Fixed missing encryption of outgoing messages when CPRIVMSG/CONTICE is used - issue #65
	* Avoid delays during DH1080 key exchange - issue #61
	* Minor bugfixes for DH1080 key exchange - issue #58
	* Updated minhook library
	* Compiled with Visual Studio 2019

== 2019-10-06 ==
	* Compatible with mIRC 7.56 and above. No longer compatible with 7.55 and below.
	* OpenSSL updated to 1.1.1d - issue 57
	* Outgoing cryptmark fixed (thx silv3rr) - issue 55
	* Rare failure of DH1080 handshakes fixed (thx maroonbells) - issue 58
	* Building OpenSSL based on https://github.com/kiyolee/openssl1_1-win-build
	* No more explicit dependency on any particular MSVC++ runtime
	* Dropped support for Windows XP and Server 2003

== 2018-04-13 ==
	* OpenSSL updated to 1.0.2o

== 2017-07-02 ==
	* Fixed compat with CAP LS changes in mIRC 7.49 (issue 50)
	* OpenSSL updated to 1.0.2l

== 2017-02-25 ==
	* Added new enforce_max_key_length option to blow.ini

== 2017-02-24 ==
	* Happy New Year
	* OpenSSL updated to 1.0.2k

== 2016-06-15 ==
	* OpenSSL updated to 1.0.2h
	* Added version to /fishdebug output
	* Do not allow setting ECB keys that exceed 56 bytes

== 2016-03-02 ==
	* OpenSSL updated to 1.0.2g

== 2015-11-08 ==
	* OpenSSL updated to 1.0.2d

== 2015-01-27 ==
	* OpenSSL updated to 1.0.2
	* Installer: added support for mIRC with mirc.exe + mirc.ini in the same
	  directory but portable flag set to no. (issue 39)

== 2015-01-08 ==
	* OpenSSL updated to 1.0.1k
	* Added hint to OpenSSL DLL error message
	* Fixed minor syntax mistake in fish_10.mrc

== 2014-11-29 ==
	* Added support for STARTTLS (new in mIRC 7.38)
	* OpenSSL updated to 1.0.1j
	* Restored working order on SSE 1-only CPUs (again)
	* This is a recommended update!

== 2014-08-25 ==
	* Switch to MinHook library. This enables Wine compatibility.
	* There's no need to upgrade to this version unless you run mIRC in Wine.

== 2014-08-07 ==
	* Improved error handling
	* Installer sets load=1 in [ssl] (mirc.ini) - required for mIRC 7.35
	* OpenSSL updated to 1.0.1i

== 2014-07-05 ==
	* Small improvements to the installer
	* New /fishdebug command, kindly provided by devn0ll!

== 2014-06-22 ==
	* Fix for mIRC 7.34 + ZNC 1.x [ CAP server-time / IRC Message Tags support ]
	* New setup routine (installer)
	* New debugging commands
	* New README
	* Misc. internal changes

== 2014-06-12 ==
	* New internal startup flow
	* fish_10.mrc file MUST be updated!
	* OpenSSL update to 1.0.1h

== 2014-04-08 ==
	* OpenSSL update to 1.0.1g

== 2014-01-11 ==
	* OpenSSL update to 1.0.1f (fixes some security issues)
	* Restored working order on SSE 1-only CPUs
	* Some internal changes to memory management

== 2013-09-23 ==
	* Support log decryption from psyBNC 2.4 [don't bother unless you use it]

== 2013-02-13 ==
	* OpenSSL update to 1.0.1e [don't bother if you are running 2013-02-10 and not experiencing SSL issues]

== 2013-02-10 ==
	* OpenSSL update to 1.0.1d
	* Recompiled main DLLs with Visual Studio 2012 - may or may not improve performance.

== 2013-01-06 ==
	* Added support for psyBNC log buffer decryption...

== 2012-November ==
	* Nothing broke so far - despite the major changes and additions, I'm a freakin' genius!

== 2012-06-25 ==
	* Rewrote injection logic, fixing the 7.25 and hopefully also all possible future issues like these.
	* Added support for IRC connections through SOCKS4, SOCKS5 and HTTP-CONNECT proxy servers.
	* A bit of a testing release, please let me know if something breaks!

== 2012-06-20 ==
	* Added a workaround for the previous incompatibility with fresh mIRC 7.25 installations, this time
	  for SSL-enabled connections. Sorry for the trouble.

== 2012-06-14 ==
	* Fixed a slight incompatibility with fresh mIRC 7.25 installations that caused decryption and
	  encryption to stop working.

== 2012-05-30 ==
	* OpenSSL update to 1.0.1c (some IRC servers were not working with 1.0.1a, thanks to who reported this)

== 2012-04-19 ==
	* OpenSSL update to 1.0.1a (preventing a possible security issue)

== 2012-01-30 ==
	* Fixed an issue identified by someone on IRC: when using sockopen -e (SSL) sockets and receiving stuff
	  before writing to the socket, the writes would always fail.
	* Updated OpenSSL to 1.0.0g

== 2011-11-17 ==
	* Enabled mIRC6-compatibility encrypted ACTION (/me) workaround to actually work (incoming messages).
	* Fixed outgoing encrypted /me ACTIONs. They were missing their last character.
	* Proper support of UTF-8 encoded cryptmarks when using no_legacy=1 mode without $chr() workaround.
	* Updated OpenSSL to 1.0.0e

== 2011-03-15 ==
	* Updated FiSH_encrypt_msg and FiSH_decrypt_msg DLL routines (for use from within .mrc scripts)
	  to be fully backwards-compatible again (keys with characters >=0x80 were broken).
	* Added FiSH_DecryptMsg10 and FiSH_EncryptMsg10 DLL routines with the following signature:
	  <decode_utf8|raw_bytes> <key> <base64data>
	  Use these if you need to work with UTF-8 encoded keys, e.g.:

src/blowini.cpp
#include "fish-internal.h"
#include <time.h>

#define INI_SECTION L"FiSH"


CBlowIni::CBlowIni(std::wstring a_iniPath)
{
	SetIniPath(a_iniPath);

	m_iniBlowKey = "blowinikey\0ADDITIONAL SPACE FOR CUSTOM BLOW.INI PASSWORD";
	// well, it's open source now, but hex editing the DLL might still
	// be easier in some cases.
}


void CBlowIni::SetIniPath(std::wstring a_iniPath)
{
	m_iniPath = a_iniPath;
	// allow no_legay setting to be read in GetBool call... *hack*:
	m_noLegacy = false;
	m_noLegacy = GetBool(L"no_legacy", false);
}


bool CBlowIni::IsWritable() const
{
	wchar_t l_buf[20] = {0};
	swprintf_s(l_buf, 19, L"%d", time(NULL));

	return (::WritePrivateProfileStringW(L"Test", L"StartUp", l_buf, m_iniPath.c_str()) != 0);
}


std::wstring CBlowIni::GetStringW(const wchar_t* a_key, const wchar_t* a_default) const
{
	wchar_t l_buf[4096] = {0};

	::GetPrivateProfileStringW(INI_SECTION, a_key, a_default, l_buf, 4095, m_iniPath.c_str());

	return l_buf;
}


std::string CBlowIni::GetString(const wchar_t* a_key, const wchar_t* a_default) const
{
	return UnicodeToCp(CP_UTF8, GetStringW(a_key, a_default));
}


bool CBlowIni::GetBool(const wchar_t* a_key, bool a_default) const
{
	const std::wstring l_val = GetStringW(a_key, (a_default ? L"1" : L"0"));

	return (l_val.empty() ? a_default : (l_val[0] != L'0' && l_val[0] != L'n' && l_val[0] != L'N'));
}


int CBlowIni::GetInt(const wchar_t* a_key, int a_default) const
{
	return ::GetPrivateProfileIntW(INI_SECTION, a_key, a_default, m_iniPath.c_str());
}


bool CBlowIni::SetInt(const wchar_t* a_key, int a_value) const
{
	wchar_t l_buf[20] = {0};
	swprintf_s(l_buf, 19, L"%d", a_value);

	return (::WritePrivateProfileStringW(INI_SECTION, a_key, l_buf, m_iniPath.c_str()) != FALSE);
}


std::string CBlowIni::FixContactName(const std::string& a_name)
{
	std::string l_result;

	l_result.reserve(a_name.size());

	for(size_t p = 0; p < a_name.size(); p++)
	{
		l_result += (a_name[p] == '[' || a_name[p] == ']' ? '~' : a_name[p]);
	}

	return l_result;
}


std::string CBlowIni::GetBlowKey(const std::string& a_name, bool& ar_cbc) const
{
	const std::string l_iniSection = FixContactName(a_name);
	std::string l_blowKey;

	if(m_noLegacy)
	{
		const std::wstring l_iniSectionW = UnicodeFromCp(CP_UTF8, l_iniSection);

		wchar_t l_buf[4096] = {0};
		::GetPrivateProfileStringW(l_iniSectionW.c_str(), L"key", L"", l_buf, 4095, m_iniPath.c_str());

		l_blowKey = UnicodeToCp(CP_UTF8, l_buf);
	}
	else
	{
		const std::string l_ansiFileName = UnicodeToCp(CP_ACP, m_iniPath);
		char l_buf[4096] = {0};

		// use ANSI to maintain backwards compatibility to old blow.ini / fish.dlls.
		::GetPrivateProfileStringA(l_iniSection.c_str(), "key", "", l_buf, 4095, l_ansiFileName.c_str());

		if(*l_buf == 0)
		{
			// fallback for INI filenames with non-ANSI characters:
			const std::wstring l_iniSectionW = UnicodeFromCp(CP_ACP, l_iniSection);

			wchar_t l_buf[4096] = {0};
			::GetPrivateProfileStringW(l_iniSectionW.c_str(), L"key", L"", l_buf, 4095, m_iniPath.c_str());

			l_blowKey = UnicodeToCp(CP_ACP, l_buf);
		}
		else
		{
			l_blowKey = l_buf;
		}
	}

	if(l_blowKey.find("+OK ") == 0)
	{
		// the blowfish_decrypt must not change in a way that makes it stop
		// making a copy of the first argument or this will blow up.
		if(blowfish_decrypt(l_blowKey.substr(4), l_blowKey, m_iniBlowKey) != 0)
		{
			l_blowKey = "";
		}
	}

	ar_cbc = (HasCBCPrefix(l_blowKey, true) || GetSectionBool(a_name, L"cbc", false));

	return l_blowKey;
}


std::string CBlowIni::GetBlowKey(const std::string& a_network, const std::string& a_contact, bool& ar_cbc) const
{
#if 0
	if(a_network.empty())
	{
		return GetBlowKey(a_contact, ar_cbc);
	}
	else
#endif
	{
		std::string l_result = GetBlowKey(a_network + ":" + a_contact, ar_cbc);

		if(!m_noLegacy && l_result.empty())
		{
			// fall back to legacy non-network name-prefixed section:
			l_result = GetBlowKey(a_contact, ar_cbc);
		}

		return l_result;
	}
}


bool CBlowIni::DeleteBlowKey(const std::string& a_name) const
{
	const std::string l_iniSection = FixContactName(a_name);
	bool l_success = true;

	if(m_noLegacy)
	{
		const std::wstring l_iniSectionW = UnicodeFromCp(CP_UTF8, l_iniSection);

		l_success = (::WritePrivateProfileStringW(l_iniSectionW.c_str(), NULL, NULL, m_iniPath.c_str()) != 0);
	}
	else
	{
		const std::string l_ansiFileName = UnicodeToCp(CP_ACP, m_iniPath);

		if(::WritePrivateProfileStringA(l_iniSection.c_str(), NULL, NULL, l_ansiFileName.c_str()) == 0)
		{
			// compatibility fallback, read above...
			const std::wstring l_iniSectionW = UnicodeFromCp(CP_ACP, l_iniSection);

			l_success = (::WritePrivateProfileStringW(l_iniSectionW.c_str(), NULL, NULL, m_iniPath.c_str()) != 0);
		}
	}

	return l_success;
}


bool CBlowIni::DeleteBlowKey(const std::string& a_network, const std::string& a_contact) const
{
	bool l_result = DeleteBlowKey(a_network + ":" + a_contact);

	if(!m_noLegacy)
	{
		l_result = DeleteBlowKey(a_contact) || l_result;
	}

	return l_result;
}


bool CBlowIni::WriteBlowKey(const std::string& a_name, const std::string& a_value) const
{
	const std::string l_iniSection = FixContactName(a_name);
	std::wstring l_iniSectionW;
	std::string l_keyActual(a_value), l_keyValue;
	bool l_cbc = HasCBCPrefix(l_keyActual, true);

	blowfish_encrypt(l_keyActual, l_keyValue, m_iniBlowKey);
	l_keyValue.insert(0, "+OK ");

	bool l_success = true;

	if(m_noLegacy)
	{
		l_iniSectionW = UnicodeFromCp(CP_UTF8, l_iniSection);
		const std::wstring l_keyValueW = UnicodeFromCp(CP_UTF8, l_keyValue);
		l_success = (::WritePrivateProfileStringW(l_iniSectionW.c_str(), L"key", l_keyValueW.c_str(), m_iniPath.c_str()) != 0);
	}
	else
	{
		const std::string l_ansiFileName = UnicodeToCp(CP_ACP, m_iniPath);
		l_iniSectionW = UnicodeFromCp(CP_ACP, l_iniSection);

		if(::WritePrivateProfileStringA(l_iniSection.c_str(), "key", l_keyValue.c_str(), l_ansiFileName.c_str()) == 0)
		{
			const std::wstring l_keyValueW = UnicodeFromCp(CP_ACP, l_keyValue);

			l_success = (::WritePrivateProfileStringW(l_iniSectionW.c_str(), L"key", l_keyValueW.c_str(), m_iniPath.c_str()) != 0);
		}
	}

	if(l_success)
	{
		wchar_t l_timeBuf[30] = {0};
		time_t l_time = time(NULL);
		struct tm l_today;
		_localtime64_s(&l_today, &l_time);
		if(wcsftime(l_timeBuf, 29, L"%d/%m/%Y", &l_today) > 0)
		{
			::WritePrivateProfileStringW(l_iniSectionW.c_str(), L"date", l_timeBuf, m_iniPath.c_str());
		}
	}

	if(l_success && l_cbc)
	{
		::WritePrivateProfileStringW(l_iniSectionW.c_str(), L"cbc", L"1", m_iniPath.c_str());
	}
	else
	{
		// delete cbc entry
		::WritePrivateProfileStringW(l_iniSectionW.c_str(), L"cbc", NULL, m_iniPath.c_str());
	}

	return l_success;
}


bool CBlowIni::WriteBlowKey(const std::string& a_network, const std::string& a_contact, const std::string& a_value) const
{
	bool l_result = WriteBlowKey(a_network + ":" + a_contact, a_value);

	if(!m_noLegacy)
	{
		// (over)write key, legacy style:
		l_result = WriteBlowKey(a_contact, a_value) || l_result;
	}

	return l_result;
}


bool CBlowIni::GetSectionBool(const std::string& a_name, const wchar_t* a_key, bool a_default) const
{
	const std::string l_iniSection = FixContactName(a_name);
	int l_result = -1;

	if(!m_noLegacy)
	{
		const std::string l_ansiFileName = UnicodeToCp(CP_ACP, m_iniPath);
		const std::string l_iniKey = UnicodeToCp(CP_ACP, a_key);
		char l_buf[10] = {0};

		::GetPrivateProfileStringA(l_iniSection.c_str(), l_iniKey.c_str(),
			(a_default ? "1" : "0"), l_buf, 9, l_ansiFileName.c_str());

		l_result = (l_buf[0] != L'0' && l_buf[0] != L'n' && l_buf[0] != L'N' ? 1 : 0);
	}

	if(m_noLegacy || l_result == -1)
	{
		const std::wstring l_iniSectionW = UnicodeFromCp((m_noLegacy ? CP_UTF8 : CP_ACP), l_iniSection);
		wchar_t l_bufW[10] = {0};

		::GetPrivateProfileStringW(l_iniSectionW.c_str(), a_key,
			(a_default ? L"1" : L"0"), l_bufW, 9, m_iniPath.c_str());

		return (!*l_bufW ? a_default : (l_bufW[0] != L'0' && l_bufW[0] != L'n' && l_bufW[0] != L'N'));
	}

	return (l_result != 0);
}


bool CBlowIni::GetSectionBool(const std::string& a_network, const std::string& a_contact, const wchar_t* a_key, bool a_default) const
{
	if(m_noLegacy)
	{
		return GetSectionBool(a_network + ":" + a_contact, a_key, a_default);
	}
	else
	{
		return GetSectionBool(a_network + ":" + a_contact, a_key, GetSectionBool(a_contact, a_key, a_default));
	}
}


bool CBlowIni::SetSectionValue(const std::string& a_name, const wchar_t* a_key, const std::wstring& a_value) const
{
	const std::string l_iniSection = FixContactName(a_name);
	std::wstring l_iniSectionW;

	bool l_success = true;

	if(m_noLegacy)
	{
		l_iniSectionW = UnicodeFromCp(CP_UTF8, l_iniSection);
		l_success = (::WritePrivateProfileStringW(l_iniSectionW.c_str(), a_key, a_value.c_str(), m_iniPath.c_str()) != 0);
	}
	else
	{
		const std::string l_ansiFileName = UnicodeToCp(CP_ACP, m_iniPath),
			l_ansiValue = UnicodeToCp(CP_ACP, a_value),
			l_ansiKey = UnicodeToCp(CP_ACP, a_key);

		l_iniSectionW = UnicodeFromCp(CP_ACP, l_iniSection);

		if(::WritePrivateProfileStringA(l_iniSection.c_str(), l_ansiKey.c_str(), l_ansiValue.c_str(), l_ansiFileName.c_str()) == 0)
		{
			l_success = (::WritePrivateProfileStringW(l_iniSectionW.c_str(), a_key, a_value.c_str(), m_iniPath.c_str()) != 0);
		}
	}

	return l_success;
}


bool CBlowIni::SetSectionBool(const std::string& a_network, const std::string& a_contact, const wchar_t* a_key, bool a_value) const
{
	wchar_t l_buf[3] = {0};
	swprintf_s(l_buf, 2, L"%d", (a_value ? 1 : 0));

	bool l_result = SetSectionValue(a_network + ":" + a_contact, a_key, l_buf);

	if(!m_noLegacy)
	{
		// (over)write, legacy style:
		l_result = SetSectionValue(a_contact, a_key, l_buf) || l_result;
	}

	return l_result;
}
